### ISSUE TITLE
Add a dotnet-ref command to enable adding references to the project file 

### ISSUE THREAD

**Issue by blackdwarf** [dotnet/cli#4521]
### Top level Verb

ref
### Sub verbs

add | del| list
### Syntax

dotnet ref &lt;operation&gt; &lt;id&gt; &lt;options&gt;
### Description

This command will add a reference to either a file system assembly, a reference assembly or another project. This will result in the mutation of the project file. If a reference to a NuGet package is added, the project will be restored afterwards, to enable immediate usage of the dependency that was added. The reference added will be added in the "general" pool of references; this command will not put TFM-specific dependencies in the project file. For that, a manual edit will have to be done. If the project file is being added and the project file specified does not exist, an error will be show. If just a path to a directory is found, the command will try to detect a project file and if a project file exists it will add it as a project reference.
### Switches
-   --type | -t
  -   Force the type of reference to add. Permitted values are "project", "package", "assembly"
-   --version | v
  -   Which version of the reference to add; applicable to NuGet packages only. If this parameter is not specified, "latest" is assumed which means a "*".
### Arguments
-   &lt;id&gt;
  -   By default, a string passed after the sub-verb will be treated as the --id switch value
### Defaults
-   If just &lt;id&gt; exists, we make an effort to “sniff” the reference out.
### Examples
-   dotnet ref add "../../ClassLib1/classlib1.csproj"
  -   Adds a project "ClassLib1" from a given path
-   dotnet ref add "../../../assembly name with a space.dll"
  -   Adds the assembly that exists on FS
-   dotnet ref rem "../../ClassLib1/classlib1.csproj"
  -   Removes the specified P2P dependency from the references
-   dotnet ref add Newtonsoft.JSON
  -   Add a &lt;PackageReference&gt; to JSON.NET
-   dotnet ref add Newtonsoft.JSON --version 9.0.1
  -   Add JSON.net package reference with a version set to 9.0.1
-   dotnet ref list
  -   Lists out all of the references; the display should be grouped by reference type with friendly names
-   dotnet ref list --type project
  -   List out only the project references

**Comment by dasMulli** [comment]
It would be awesome to also have a sub-verb to update NuGet references / show that an update is available.

**Comment by wjk** [comment]
Honestly, I don't see a need for a specialized command just to add a `<PackageReference>` tag to the csproj; I do it by hand and have no problems. On the other hand, @dasMulli's suggestion about a command to update NuGet references is a _great_ idea.

**Comment by Ziflin** [comment]
@wjk It is a very useful feature of NodeJS to be able to install/uninstall a package from the command line for certain build systems. I would imagine it would be equally useful in dotnet.

**Comment by dasMulli** [comment]
> this command will not put TFM-specific dependencies in the project file. For that, a manual edit will have to be done

I think if it did support TFM-specific references, this would be highly useful since adding a PackageReference is easy, but getting the condition right from memory not so much (srsly, "do i need to wrap things with apostrophes? do i need brackets for the `$xyz` expressions? what was that variable called again anyway?")
 
For adding references, a switch could be added for the target framework:

* `--framework | -v`
Specify for which target framework the reference shall be added to.

Use cases:
* `dotnet ref add Microsoft.AspNetCore.Hosting.WindowsServices -f net462`
* `dotnet ref add Microsoft.NETCore.App -v 1.1.0 -f netcoreapp1.1`

Logic could be (design needed):
1.  If the package is already in a general ItemGroup, fail with an error.
2. If there is a PackageReference item with the required condition in a general ItemGroup, fail with an error.
3. If the package is already in an item group with a TFM condition on the specified framework, fail with an error.
4. Find or create an ItemGroup with the necessary TFM condition, add a PackageReference there.

If VS "15", NuGet 4.* VSIX already has logic for that, it could be "borrowed".

(Same applies to projects / ProjectReference)

**Comment by blackdwarf** [comment]
Tracking issue is #4746

**Comment by enricosada** [comment]
It's possible to add also an option to set also `PrivateAssets` property for `PackageReference`?
it's useful for adding sdk packages who need `<PrivateAssets>All</PrivateAssets>`

So something like `--private-assets "All"` ? 
@blackdwarf  if ok, i can add that (i need that for adding fsharp sdk package :D, so personal interests) but i think is useful for scripting/tutorial to add `dotnet test` sdk packages
So like

- create console with `dotnet new`
- add ref to xunit/nunit sdk
- add ref to xunit/nunit packages
- go!

**Comment by piotrpMSFT** [comment]
@enricosada I like the proposal. Can you wait on #4758? We decided to change the calling convention for this class of commands to get back to `dotnet {verb} {subject} {object}` behavior and @krwq is making fairly big code changes to account for that. Once it's merged, we'll have a basis for other commands including adding package refs and the private flag.

@rrelyea I believe this ask, having a verb to add a nuget package reference and marking it <PrivateAssets>All<... might be interesting to you.

**Comment by piotrpMSFT** [comment]
@dasMulli Can you open an issue in nuget/home to track the PackageRef update verb you propose?

**Comment by piotrpMSFT** [comment]
@blackdwarf can you update the docs at the top to match the current POR for impl?

**Comment by enricosada** [comment]
> Honestly, I don't see a need for a specialized command just to add a <PackageReference> tag to the csproj; I do it by hand and have no problems

the command is for cli, but is also useful because you can automate things (from ide/tools or script). 

- dotnet new
- dotnet ref add Newtonsoft.Json
- copy Program.fs serializator example
- dotnet run

**Comment by krwq** [comment]
`dotnet add p2p` `dotnet remove p2p` and `dotnet list p2ps` are already here. Dlls are out of scope. Packages are done by NuGet team.

@piotrpMSFT, can we close this? We already have https://github.com/dotnet/cli/issues/4746

**Comment by piotrpMSFT** [comment]
Yea, dotnet add p2p is done.

@enricosada make sure the add ref features you're asking for get logged in nuget/home.
